### ISSUE TITLE
Suggestion: disallow additional items in tuples

### ISSUE THREAD

**Issue by 43081j** [lukeed/tschema#7]
Currently, a tuple will produce something like this:

```ts
const output = t.tuple([
  t.number()
]);

// output is the following:
{
  type: 'array',
  prefixItems: [{type: "number"}]
}
```

this will successfully validate against the following:

```json
[
  100,
  "foo"
]
```

similar to the discussion around strict objects, this aligns with json schema's defaults but feels confusing on the JS side of things

maybe we should default to `items: false`?

**Comment by lukeed** [comment]
I was going to make an issue about this too, as tuples are the weirdest part of JSON schema.

As you pointed out, you can add extra items by default (see [spec](https://json-schema.org/understanding-json-schema/reference/array#additionalitems)). But according to JSON schema, you also don't even have to provide all defined items (see [spec, 2nd to last example](https://json-schema.org/understanding-json-schema/reference/array#tupleValidation)).

To address the first point, the current `t.tuple()` definition accepts the `{ items: false }` option (and **only** `false`).
But there's nothing preventing a user from omitting some of the trailing tuple members (except TS error).

I don't have an answer re: what to do yet, but it's a balance between respecting JSON schema (primary lib purpose) and generating definitions that match JS expectations.

In other words, all that makes a "tuple" is the use of `prefixItems` instead of `items` and, sure, adding `items: false` by default still preserves an avenue for users to say "i want additional items" (via `items` or `unevaluatedItems` definition). This defines a valid schema and, let's say, the `[string, number, boolean]` TS type. But then when validating, to remain JSON spec compliant, I have to still pass validation when only `[string, number]` comes in, despite it failing the TS check and the "JS expectation" gut-check.

**Comment by 43081j** [comment]
yeah that is far from ideal, i hadn't realised you can also pass fewer items than are meant to be in the tuple. that seems the opposite of what most languages with tuples would allow

maybe it is worth doing a poll some place to find out if people would expect the JS behaviour or the schema behaviour

**Comment by lukeed** [comment]
I'm afraid spec compliance is a must-have / non-negotiable. The beauty of JSON schema is that its language agnostic and works everywhere. The (wip) validator can have an option that purposefully breaks out of that tuple strangeness, but by default it needs to abide by spec'd & shared behavior

**Comment by 43081j** [comment]
i was more referring to the default being `items: false` or not, FYI

does the spec explicitly define what the defaults should be? and if it does, what do you think needs to be decided then? if that's the case, it has to be the way it is right now afaict

**Comment by lukeed** [comment]
This can be solved with:

```ts
{
  // ...
  additionalItems: false // prevent extras (recent spec drafts only)
  maxItems: 4, // example, length of tuple input args
  minItems: 4, // length of tuple input args
}
```

> https://www.jsonschemavalidator.net/s/rIPUTEE4

The `additionalItems` and `maxItems` overlap responsibility and are technically redundant (in eyes of newer specs only)

**Comment by lukeed** [comment]
While arguably not breaking, going to include this as part of the 3.0 major release (spearheaded by #5)
